boards/PDM/Src/Io/Io_Efuse.c

```diff
@@ -0,0 +1,935 @@
+    command = (uint16_t)(command | (register_value & EFUSE_SI_DATA_MASK));
+    // Invert watchdog bit state (Note: It is safe to do so even if the watchdog
+    // is disabled)
+    if (e_fuse->wdin_bit_to_set)
+    {
+        SET_BIT(command, WATCHDOG_BIT);
+    }
+    else
+    {
+        CLEAR_BIT(command, WATCHDOG_BIT);
+    }
+    // Invert watchdog bit state for next write
+    e_fuse->wdin_bit_to_set = !e_fuse->wdin_bit_to_set;
+
+    // Compute and set/clear parity value
+    Io_Efuse_CalculateParityBit(&command);
+
+    exit_code =
+        Io_Efuse_WriteToEfuse(&command, chip_select_port, chip_select_pin);
+
+    return exit_code;
+}
+
+static ExitCode Io_Efuse_ReadRegister(
+    uint8_t         register_address,
+    uint16_t *      register_value,
+    GPIO_TypeDef *  chip_select_port,
+    uint16_t        chip_select_pin,
+    struct EfuseIo *e_fuse)
+{
+    ExitCode exit_code;
+    uint16_t command = 0x0000U;
+
+    // Place the Status Register address into bits 10->13
+    command =
+        (uint16_t)((SI_STATR_0_ADDR & EFUSE_ADDR_MASK) << EFUSE_ADDR_SHIFT);
+    // Shift bit 3 of the address (SOA3: the channel number) to the 13th bit
+    command = (uint16_t)(
+        command | ((register_address & SOA3_MASK) << EFUSE_ADDR_SHIFT) |
+        (register_address & (SOA2_MASK | SOA1_MASK | SOA0_MASK)));
+    // Invert watchdog bit state (Note: It is safe to do so even if the watchdog
+    // is disabled)
+    if (e_fuse->wdin_bit_to_set)
+    {
+        SET_BIT(command, WATCHDOG_BIT);
+    }
+    else
+    {
+        CLEAR_BIT(command, WATCHDOG_BIT);
+    }
+    // Invert watchdog bit state for next write
+    e_fuse->wdin_bit_to_set = !e_fuse->wdin_bit_to_set;
+
+    // Compute and set/clear parity bit
+    Io_Efuse_CalculateParityBit(&command);
+
+    exit_code = Io_Efuse_ReadFromEfuse(
+        &command, register_value, chip_select_port, chip_select_pin);
+    // Only return register contents and clear bits 9->15
+    CLEAR_BIT(*register_value, ~EFUSE_SO_DATA_MASK);
+
+    return exit_code;
+}
+
+static void Io_Efuse_CalculateParityBit(uint16_t *spi_command)
+{
+    uint16_t spi_data = *spi_command;
+    bool     parity_bit;
+
+    for (parity_bit = 0; spi_data > 0; spi_data >>= 1)
+    {
+        parity_bit ^= READ_BIT(spi_data, 1U);
+    }
+
+    // Set the parity bit if it should be set, else
+    // clear the bit
+    if (parity_bit)
+    {
+        SET_BIT(*spi_command, PARITY_BIT);
+    }
+    else
+    {
+        CLEAR_BIT(*spi_command, PARITY_BIT);
+    }
+}
+
+static ExitCode Io_Efuse_WriteToEfuse(
+    uint16_t *    tx_data,
+    GPIO_TypeDef *chip_select_port,
+    uint16_t      chip_select_pin)
+{
+    HAL_StatusTypeDef status;
+
+    HAL_GPIO_WritePin(chip_select_port, chip_select_pin, GPIO_PIN_RESET);
+    status = HAL_SPI_TransmitReceive(
+        efuse_spi_handle, (uint8_t *)tx_data, (uint8_t *)tx_data, 1U, 100U);
+    HAL_GPIO_WritePin(chip_select_port, chip_select_pin, GPIO_PIN_SET);
```

Comment:
```suggestion
    HAL_GPIO_WritePin(chip_select_port, chip_select_pin, GPIO_PIN_RESET);
    HAL_StatusTypeDef status = HAL_SPI_TransmitReceive(
        efuse_spi_handle, (uint8_t *)tx_data, (uint8_t *)tx_data, 1U, 100U);
    HAL_GPIO_WritePin(chip_select_port, chip_select_pin, GPIO_PIN_SET);
```

Reply:
Do we not normally declare local variables at the beginning of a block?

Reply:
Idk if this was enforced. Just trying to save some lines

Reply:
I see DIM project follows the convention for local variables. I should initialize it though for consistency.

Reply:
I think we are far from saving lines at this point lol